src/net/socket_ancillary.rs

```diff
@@ -0,0 +1,260 @@
+pub struct SocketCred(c::ucred);
+
+/// TODO: document
+pub struct AncillaryDataIter<'a, T> {
+    data: &'a [u8],
+    _t: PhantomData<T>,
+}
+
+/// A type-safe zero-copy wrapper around a list of [`SendControlMessage`s].
+pub struct SendControlMessages<'a> {
+    bytes: &'a [u8],
+}
+
+// TODO: Provide way of sizing the buffer for SocketAncillary upfront, like in
+// https://docs.rs/nix/latest/nix/macro.cmsg_space.html
+
+// TODO: Find a way to use MaybeUninit as backing data.
+
+// TODO: Should there exist a convenience wrapper that owns the buffer and potentially
+// auto resizes?
+
+// TODO: port tests from https://github.com/nix-rust/nix/blob/master/test/sys/test_socket.rs
+
+/// TODO: document
+#[derive(Debug)]
+pub struct SocketAncillary<'a> {
+    buffer: &'a mut [u8],
+    length: usize,
+    truncated: bool,
+}
+
+impl<'a> SocketAncillary<'a> {
+    /// Create an ancillary data with the given buffer.
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        SocketAncillary {
+            buffer,
+            length: 0,
+            truncated: false,
+        }
+    }
+
+    /// Returns the capacity of the buffer.
+    pub fn capacity(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns `true` if the ancillary data is empty.
+    pub fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    /// Returns the number of used bytes.
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    /// Returns the iterator of the control messages.
+    pub fn messages(&self) -> Messages<'_> {
+        Messages {
+            buffer: &self.buffer[..self.length],
+            current: None,
+        }
+    }
+
+    /// Is `true` if during a recv operation the ancillary was truncated.
+    pub fn truncated(&self) -> bool {
+        self.truncated
+    }
+
+    /// Clears the ancillary data, removing all values.
+    pub fn clear(&mut self) {
+        self.length = 0;
+        self.truncated = false;
+    }
+}
+
+/// TODO: document
+#[derive(Debug)]
+pub struct UnixSocketAncillary<'a>(SocketAncillary<'a>);
+
+impl<'a> Deref for UnixSocketAncillary<'a> {
+    type Target = SocketAncillary<'a>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'a> DerefMut for UnixSocketAncillary<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<'a> UnixSocketAncillary<'a> {
+    /// Create an ancillary data with the given buffer.
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        UnixSocketAncillary(SocketAncillary::new(buffer))
+    }
+
+    /// Add file descriptors to the ancillary data.
+    ///
+    /// The function returns `true` if there was enough space in the buffer.
+    /// If there was not enough space then no file descriptors was appended.
+    /// Technically, that means this operation adds a control message with the level `SOL_SOCKET`
+    /// and type `SCM_RIGHTS`.
+    pub fn add_fds<Fd: AsRawFd>(&mut self, fds: &[Fd]) -> bool {
```

Comment:
Would it be feasible to use `AsFd` here, instead of `AsRawFd`? You can still get the raw fds via `as_fd().as_raw_fd()`.

Under I/O safety, if this takes `AsRawFd`, it should be marked `unsafe`.

Reply:
Should be able to change that, yes.

Reply:
fixed